File: src/Functional/Elixir/facade.exs
defmodule AuthService do
  def authenticate(user), do: "auth(#{user})"
end

defmodule InventoryService do
  def reserve(sku), do: "reserve(#{sku})"
end

defmodule BillingService do
  def charge(cents), do: "charge(#{cents})"
end

defmodule CheckoutFacade do
  def checkout(user, sku, cents) do
    [
      AuthService.authenticate(user),
      InventoryService.reserve(sku),
      BillingService.charge(cents)
    ]
    |> Enum.join(">")
  end
end

IO.puts("checkout=#{CheckoutFacade.checkout("alice", "SKU-42", 499)}")

File: src/Functional/Haskell/Facade.hs
module Main where

authenticate :: String -> String
authenticate user = "auth(" ++ user ++ ")"

reserve :: String -> String
reserve sku = "reserve(" ++ sku ++ ")"

charge :: Int -> String
charge cents = "charge(" ++ show cents ++ ")"

checkout :: String -> String -> Int -> String
checkout user sku cents =
    "checkout=" ++ authenticate user ++ ">" ++ reserve sku ++ ">" ++ charge cents

main :: IO ()
main = putStrLn (checkout "alice" "SKU-42" 499)
